R/grid.R

```diff
@@ -0,0 +1,84 @@
+#' Create a grid of viewers that share a WebGL canvas
+#'
+#' @param viewer a list contains sub-viewers.
+#' @param element_id HTML string identifier.
+#' @param configs grid configuration.
+#' @param viewer_config viewer specification to apply to all subviewers.
+#' @param width Fixed width for viewer (in css units). Ignored when used in a
+#' Shiny app -- use the \code{width} parameter in
+#' \code{\link[r3dmol]{r3dmolOutput}}.
+#' It is not recommended to use this parameter because the widget knows how to
+#' adjust its width automatically.
+#' @param height Fixed height for viewer (in css units). It is recommended to
+#' not use this parameter since the widget knows how to adjust its height
+#' automatically.
+#'
+#' @return
+#' @export
+#'
+#' @examples
+#' library(r3dmol)
+#'
+#' m1 <- r3dmol() %>%
+#'   m_add_model(data = pdb_6zsl, format = "pdb") %>%
+#'   m_zoom_to()
+#'
+#' m2 <- m1 %>%
+#'   m_set_style(style = m_style_cartoon(color = "spectrum"))
+#'
+#' m3 <- m1 %>%
+#'   m_set_style(style = m_style_stick())
+#'
+#' m4 <- m1 %>%
+#'   m_set_style(style = m_style_sphere())
+#'
+#' m_grid(
+#'   viewer = list(m1, m2, m3, m4),
+#'   configs = list(rows = 2, cols = 2, control_all = TRUE),
+#'   viewer_config = list(backgroundColor = "black")
+#' )
+m_grid <- function(viewer,
+                   element_id,
+                   configs,
+                   viewer_config = list(),
+                   width = NULL,
+                   height = NULL) {
+  if (missing(element_id)) {
+    element_id <- (sample(256, 10, replace = TRUE) - 1) %>%
+      as.hexmode() %>%
+      format(width = 2) %>%
+      paste(collapse = "")
+  }
+  if (missing(configs)) {
+    configs <- list(
+      rows = 1,
+      cols = length(viewer),
+      control_all = TRUE
+    )
+  }
```

Comment:
I really like this implementation. The only thing I would change is maybe exposing the `viewer_config` options as their own options? As far as I can tell it's only `rows`, `cols` & `control_all` that are the potential options, and this way the user would be prompted about them and see the options in the documentation.

i.e.
```r
m_grid <- function(viewer,
                   element_id,
                   configs,
                   rows = 1,
                   cols = 1,
                   control_all = TRUE,
                   width = NULL,
                   height = NULL) {

```

Just for a the user without having to dig into the documentation. What do you think?

Reply:
Thanks! You are right that exposing `rows`, `cols` and `control_all` to user is a friendly implementation.
Another thing that can be done is to automatically set the values of `rows` and `cols` according to the length of viewers instead of providing 1,1 as a default option.

If you also think it's necessary, would you mind to help me implement the algorithm? Because there are other issues to investigate such as adding feature to export as image (#26) and buttons (#23).
Feel free to edit this PR.

Reply:
Yep will do! Should have time to do it either this evening or tomorrow morning.